src/ekore/anomalous_dimensions/unpolarized/time_like/as2.py

```diff
@@ -7,12 +7,14 @@
 from eko import constants
 from eko.constants import zeta2, zeta3
 
-from ....harmonics import w1, w2, w3
+from ....harmonics import cache as c
+
+# TODO: CLEAN: all the calls to polygamma could be reduced to some harmonics...
 from ....harmonics.polygamma import cern_polygamma as polygamma
```

Comment:
@t7phy I noticed here there are still some explicit calls to `cern_polygamma`, 
would you mind simplify them in terms of the harmonics so we can use the cache?

Reply:
I had a look at the two functions which use them, gamma_nsp and gamma_nsm, and 4 expressions caught my attention: SPMOM, SLV, SSTR2P and SSTR3P. SPMOM is equivalent to g3 (as far as I can tell). However, I am unable to identify SLV, SSTR2P and SSTR3P, and which functions to correspond to from the harmonics. 
The other problem I am having is that for some reason running pytest for functions with the new cache is giving me Fatal Python Error: Aborted. So the new cache (and therefore eko) is basically not working on my machine which is of course making things difficult for me to do even trial and error to figure out the exact functions.

Reply:
If the time-like expressions follows the same syntax as the unpolarised,  I'd say the equivalences are:
```
SPMOM -> g3
```
then you can use this rules to simplify the others in terms of harmonics and g3:
```
HarmonicRules={
 PolyGamma[0, n__] -> S[1, - 1 + n] - EulerGamma,
 PolyGamma[1, n__] -> - S[2, - 1 + n] + Zeta[2],
 PolyGamma[2, n__] -> 2 S[3, - 1 + n] - 2 Zeta[3],
 PolyGamma[3, n__] -> - 6 S[4, - 1 + n] + 6 Zeta[4],
 PolyGamma[4, n__] -> 24 S[5, - 1 + n] - 24 Zeta[5],
 PolyGamma[5, n__] -> - 120 S[6, - 1 + n] + 120 Zeta[6],
 PolyGamma[6, n__] -> 720 S[7, - 1 + n] - 720 Zeta[7]
};
``` 
ending up in: 
```
SSTR3P ->  S[3, N/2]
SSTR2P ->  S[2, N/2]
SSCHLP ->  g3 + S1/N^2 - 1/12 \[Pi]^2 (2/(N + N^2) + S[1, 1/2 + N/2] - S[1, N/2]) - 5 Zeta[3]/8
```
If you find the same and test are passing we can commit this. 
Note that all the terms with g3 should appear inside `S_{-2,1}` but we know this is not straightforward since there are alle the difference `(-1)^n` prescriptions around, so we can give up on this...

Regarding the compilation error, I suspect numba is messing up something. But `pytest` should always work since it runs with numba disabled (as you can see below ... ). I will try to investigate  tomorrow if this implementation is actually able to run with `nb.jit`

Reply:
couple things: 
1. for SSCHLP, if we have S[1, 1/2 + N/2], I guess we will need to add this to cache also right, maybe as S1p3h ..?
2. I modified SPMOM, SSTR3P, SSTR2P as above, however this is giving errors in tests: https://github.com/NNPDF/eko/actions/runs/4855646050/jobs/8654404902
3. I don't think the error is coming from numba, I had a quick look at the error text and it didnt seem so...

Reply:
hi, 
 1. Yes we are already caching S[1, 1/2 + N/2] as `S1ph`, so no problem.
 2. This errors suggest the replacement was not fully successfull. Can you open a PR?
 3. These errors are not coming from numba, I meant the one about segmentation fault, you mentioned before. 
 Do you have any more detail?

Reply:
As of now, on my laptop, pytest is just hanging after 'Collected ** items'. It's working fine on master but not on this branch.

Reply:
Just to say that for me it is running without problems ...

Reply:
> PS: something you can always try is to remove the `__pycache__` folders and recompile

Note that I'm discussing this also a bit with the numba developers https://github.com/numba/numba/issues/8926, so far without large movements ...

the upshot is to delete the "correct" pycache - which should be the one containing the *first* numba call, which for us is `quad_ker` (and friends), so `eko/ev_op/__pycache__`

Reply:
but pytest doesn't use the compiled version of the functions, right?

Reply:
> but pytest doesn't use the compiled version of the functions, right?

No it should not - numba should be deactivated there

Reply:
I fear we do have some problems... I tried to run the pegasus benchmarks (`python pegasus_bench.py`), 
first time it took a while but all good. Now if I rerun I get: 

```
 ──────────────────────────────────────────── 
  Theories: 2 OCards: 1 PDFs: 1 ext: pegasus  
 ──────────────────────────────────────────── 
Computing for theory=af2df25, ocard=cd6c3a9 and pdf=ToyLH ...
ThresholdsAtlas [0.00e+00 - 4.00e+00 - 2.02e+01 - 3.00e+04 - inf], ref=None @ None
Interpolation: number of points = 49, polynomial degree = 4, logarithmic = True
ThresholdsAtlas [0.00e+00 - 4.00e+00 - 2.02e+01 - 3.00e+04 - inf], ref=2.0000000000000004 @ 3
ThresholdsAtlas [0.00e+00 - 2.00e+00 - 1.01e+01 - 1.50e+04 - inf], ref=2.0000000000000004 @ 3
Strong Coupling: a_s(µ_R^2=2.000000)^(nf=3)=0.027852=0.350000/(4π)
{'polarized': False}
{'time_like': False}
Prepare threshold operator
Evolution: computing operators 2.000000e+00 -> 4.000000e+00, nf=3
Evolution: µ_R^2 distance: 1.000000e+00 -> 2.000000e+00
Scale Variation: (µ_F/µ_R)^2 = 5.000000e-01, mode: exponentiated
Evolution: a_s distance: 3.536335e-02 -> 2.785212e-02
Evolution: order: (2, 0), solution strategy: iterate-exact
zsh: segmentation fault  python pegasus_bench.py
```
ad disabling numba it works again:

```
(eko_dev) giacomomagni@Giacomos-MacBook-Pro benchmarks % export NUMBA_DISABLE_JIT=1
(eko_dev) giacomomagni@Giacomos-MacBook-Pro benchmarks % python pegasus_bench.py   
 ──────────────────────────────────────────── 
  Theories: 2 OCards: 1 PDFs: 1 ext: pegasus  
 ──────────────────────────────────────────── 
Computing for theory=af2df25, ocard=cd6c3a9 and pdf=ToyLH ...
ThresholdsAtlas [0.00e+00 - 4.00e+00 - 2.02e+01 - 3.00e+04 - inf], ref=None @ None
Interpolation: number of points = 49, polynomial degree = 4, logarithmic = True
ThresholdsAtlas [0.00e+00 - 4.00e+00 - 2.02e+01 - 3.00e+04 - inf], ref=2.0000000000000004 @ 3
ThresholdsAtlas [0.00e+00 - 2.00e+00 - 1.01e+01 - 1.50e+04 - inf], ref=2.0000000000000004 @ 3
Strong Coupling: a_s(µ_R^2=2.000000)^(nf=3)=0.027852=0.350000/(4π)
{'polarized': False}
{'time_like': False}
Prepare threshold operator
Evolution: computing operators 2.000000e+00 -> 4.000000e+00, nf=3
Evolution: µ_R^2 distance: 1.000000e+00 -> 2.000000e+00
Scale Variation: (µ_F/µ_R)^2 = 5.000000e-01, mode: exponentiated
Evolution: a_s distance: 3.536335e-02 -> 2.785212e-02
Evolution: order: (2, 0), solution strategy: iterate-exact
Evolution: computing operators - 1/49 took: 35.609883 s
...
``` 
 
I also verified that if I remove the `__pycache__` then I'm also able to rerun with numba (modulo compilation time)
 Any idea?

Reply:
also in the CI we did not converge to anything ... I started a new run here https://github.com/NNPDF/eko/actions/runs/4861110401/jobs/8665750747 but I don't expect anything ...

Reply:
What can it be? The optional argument of `cache.get` ? 
But we do have other functions using optional arguments and numba decorated?

Reply:
From https://numba.pydata.org/numba-doc/latest/developer/caching.html :

```
Requirements for cacheable function:

The LLVM module must be self-contained, meaning that it cannot rely on other compiled units without linking to them.

The only allowed external symbols are from the [NRT](https://numba.pydata.org/numba-doc/latest/developer/numba-runtime.html#arch-numba-runtime) or other common symbols from system libraries (i.e. libc and libm).

Debugging note:

Look for the usage of inttoptr in the LLVM IR or target_context.add_dynamic_add() in the lowering code in Python. They indicate potential usage of runtime address. Not all uses are problematic and some are necessary. Only the conversion of constant integers into pointers will affect caching.

Misuse of dynamic address or dynamic symbols will likely result in a segfault.

Linking order matters because unused symbols are dropped after linking. Linking should start from the leaf nodes of the dependency graph.
```

Reply:
Okay the problem are the recursive calls of get inside get.

Reply:
> Okay the problem are the recursive calls of get inside get.

good catch! can we do a two step process? like a top-level `get` which in turn is calling `compute_composed` for the composed sums?

Reply:
So: an infinite recursion can also lead to a segmentation fault, but on a first glance we're not falling into this scenario ... can you maybe try to investigate on a smaller code base?

Reply:
@giacomomagni another line of investigation: can you make a minimal example, in a single script, with as little code as possible, that is generating the issue?

Reply:
here is how I find it:
This script works (many times) if you compute S1 only (having commented all the `elif` containing nested calls to `get` in `src/harmonics/cache.py` ), but it fails as one of these `elif` are included for instance when you ask for `g3`: 

```
import numba as nb
from ekore.harmonics import cache as c



@nb.njit(cache=True)
def my_func(cache):
    n = 1.0
    S1 = c.get(c.S1, cache, n)
    # g3 = c.get(c.g3, cache, n)
    print(S1)

cache = c.reset()
my_func(cache)
```
Of course each time you change something in `src/harmonics/cache.py` but not in the snippet you need to remove the `__pycache__` manually.
To have more insight of  what happens you can also do `export NUMBA_DEBUG_CACHE=1`

Reply:
> So: an infinite recursion can also lead to a segmentation fault, but on a first glance we're not falling into this scenario ... can you maybe try to investigate on a smaller code base?

The problem is not the infinite recursion (if that was the issue you will not even able to run anything with `export NUMBA_DISABLE_JIT=1`), but I believe is how numba deals with the `__pycache__`

Reply:
> This script works (many times) if you compute S1 only (having commented all the `elif` containing nested calls to `get` in `src/harmonics/cache.py` ), but it fails as one of these `elif` are included for instance when you ask for `g3`:

Thanks for the script, but if possible an even more minimal version would be helpful.
Can you replace the `harmonics` import with the actual cache, and trim it down until you have something minimal that generates the issue?
I.e., it would be really helpful to have a minimal *self-contained* failing script (no imports but Numba).

I do not believe that you need the whole `ekore` to make it crash (otherwise it could only be something like a memory issue), so it would make sense to make it more and more minimal. Your script is a great starting point, but taking into account `ekore` still too complex.

Reply:
Sure here you have:

```
import numba as nb
import numpy as np

# fake functions
@nb.njit(cache=True)
def f1():
    return 1
@nb.njit(cache=True)
def f2(x):
    return 2 + x

# fake keys
F1 = 0
F2 = 1

# reset
@nb.njit(cache=True)
def reset():
    return np.full(2, np.nan, np.float_)

@nb.njit(cache=True)
def not_working_get(key, cache):
    s = cache[key]
    if not np.isnan(s):
        return s 
    if key == F1:
        s = f1()
    elif key == F2:
        s = f2(not_working_get(F1, cache))
    cache[key] = s
    return s

@nb.njit(cache=True)
def working_get(key, cache):
    s = cache[key]
    if not np.isnan(s):
        return s
    if key == F1:
        s = f1()
    elif key == F2:
        s = f2(f1())
    cache[key] = s
    return s

cache = reset()
# this will work
working_get(F2, cache)
# this will work once
# not_working_get(F2, cache)
print(cache[F2])
```

In any case the most important proof is [here](https://github.com/NNPDF/eko/actions/runs/4863651845/jobs/8671636590)

Reply:
> In any case the most important proof is [here](https://github.com/NNPDF/eko/actions/runs/4863651845/jobs/8671636590)

Sorry, maybe I didn't properly express myself. I didn't want a proof that there is an issue (I fully believe you), just a minimal example to play with and try to understand what's going wrong and why.

And sorry to be a bit dense, but what does exactly mean that "this will work only once"? How can I make a sensible failing `assert`?

Reply:
"this will work only once" means: you run it, it works. You run it again without modifying anything you get seg fault.
You delete `__pycache__` and run it, it will work again.

Reply:
I see, now I can investigate a bit more

Reply:
Just notice that, in the script above, I'm not even invoking recursion, because I'm directly calling it with `exit = True`.

The presence of recursion breaks the cache on its own.

Reply:
Recursion in Numba has limited supported, but for what I found at the moment, the main problem is mutual recursion. No mention found yet about cache.
https://numba.readthedocs.io/en/stable/proposals/typing_recursion.html

My best proposal is to **avoid caching the cache getter**, i.e. the recursive function.

Reply:
thanks for the reference, unfortunately it's only a draft, sigh